autorest/adal/retry.go

```diff
@@ -0,0 +1,180 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package adal
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"math"
+	"net/http"
+	"strconv"
+	"time"
+)
+
+var (
+	// statusCodesForRetry are a defined group of status code for which the client will retry
+	statusCodesForRetry = []int{
+		http.StatusRequestTimeout,      // 408
+		http.StatusTooManyRequests,     // 429
+		http.StatusInternalServerError, // 500
+		http.StatusBadGateway,          // 502
+		http.StatusServiceUnavailable,  // 503
+		http.StatusGatewayTimeout,      // 504
+	}
+)
+
+// retriableRequest provides facilities for retrying an HTTP request.
+type retriableRequest struct {
+	req *http.Request
+	rc  io.ReadCloser
+	br  *bytes.Reader
+}
+
+// prepare signals that the request is about to be sent.
+func (rr *retriableRequest) prepare() (err error) {
+	// preserve the request body; this is to support retry logic as
+	// the underlying transport will always close the reqeust body
+	if rr.req.Body != nil {
+		if rr.rc != nil {
+			rr.req.Body = rr.rc
+		} else if rr.br != nil {
+			_, err = rr.br.Seek(0, io.SeekStart)
+			rr.req.Body = ioutil.NopCloser(rr.br)
+		}
+		if err != nil {
+			return err
+		}
+		if rr.req.GetBody != nil {
+			// this will allow us to preserve the body without having to
+			// make a copy.  note we need to do this on each iteration
+			rr.rc, err = rr.req.GetBody()
+			if err != nil {
+				return err
+			}
+		} else if rr.br == nil {
+			// fall back to making a copy (only do this once)
+			err = rr.prepareFromByteReader()
+		}
+	}
+	return err
+}
+
+func removeRequestBody(req *http.Request) {
+	req.Body = nil
+	req.GetBody = nil
+	req.ContentLength = 0
+}
+
+// newRetriableRequest returns a wrapper around an HTTP request that support retry logic.
+func newRetriableRequest(req *http.Request) *retriableRequest {
+	return &retriableRequest{req: req}
+}
+
+// request returns the wrapped HTTP request.
+func (rr *retriableRequest) request() *http.Request {
+	return rr.req
+}
+
+func (rr *retriableRequest) prepareFromByteReader() (err error) {
+	// fall back to making a copy (only do this once)
+	b := []byte{}
+	if rr.req.ContentLength > 0 {
+		b = make([]byte, rr.req.ContentLength)
+		_, err = io.ReadFull(rr.req.Body, b)
+		if err != nil {
+			return err
+		}
+	} else {
+		b, err = ioutil.ReadAll(rr.req.Body)
+		if err != nil {
+			return err
+		}
+	}
+	rr.br = bytes.NewReader(b)
+	rr.req.Body = ioutil.NopCloser(rr.br)
+	return err
+}
+
+func retry(req *http.Request, backoff time.Duration, attempts int) (resp *http.Response, err error) {
+	s := http.Client{}
+	rr := newRetriableRequest(req)
```

Comment:
>(req) [](start = 26, length = 5)

just checking, is this mantaining the initials request context ?

Reply:
No

Reply:
Why not?